Add callback to allow instrumenting discarded count, and method get estimated buffer size

mezz opened the pull request:
This PR allows users of GenStage to get the estimated buffer size, and to optionally replace the log when items are discarded from a full buffer.

Currently, when items are discarded, GenStage will log 
`GenStage producer ~tp has discarded ~tp events from buffer`
This behavior is left unchanged if the new optional callback is not implemented.

By implementing the optional callback, users have the option to replace that log. In our case, we are using it to emit metrics on the number discarded to an external service, which can track the number over time and alert someone in exceptional situations.

josevalim commented on lib/gen_stage.ex:
```diff
@@ -974,6 +976,14 @@ defmodule GenStage do
+  It receives the number of excess (discarded) items from this invocation.
+  When this callback is implemented, the error log for discarded items is bypassed so that the callback can handle it.
+  """
+  @callback handle_discarded(discarded :: non_neg_integer, state :: term) :: none
```
Not a big fan of the `handle_discarded` name, since it is not an usual callback. There is a similar-ish callback in GenServer that they call `format_status` but I am not sure if that's better.

mezz replied:
I am new to naming conventions in elixir and open to ideas.
I modeled the name of this one after the other GenStage callbacks.
GenStage also has a `format_status` callback that closely matches GenServer's.

josevalim replied:
What about `format_discarded` which returns a boolean if GenStage should log or not?

mezz replied:
That sounds good to me, I'll do that.

josevalim commented on lib/gen_stage.ex:
```diff
@@ -1746,6 +1765,7 @@ defmodule GenStage do
         type: :producer,
         buffer: Buffer.new(buffer_size),
         buffer_keep: buffer_keep,
+        mod_handle_discarded: function_exported?(mod, :handle_discarded, 2),
```
We don't need to cache it. `function_exported?` checks are performance, we can simply check it before calling.

mezz replied:
Right, this was only intended as a (premature) performance optimization. I'll remove it.

mezz commented:
Thank you for the review @josevalim !
I have added `format_discarded` with the ability to control whether the original log is printed or not.

In addition, I have added the ability for `format_discarded` to return a new state. When implementing this, I ran into what I think is a bug, fixed in the commit for this PR here: https://github.com/elixir-lang/gen_stage/pull/257/commits/13311ff23c13cd74cae06610ddcbffc9aeddf84b

I have added tests using this, but please let me know if there is an issue with my change.

josevalim commented on lib/gen_stage.ex:
```diff
@@ -397,7 +397,8 @@ defmodule GenStage do
   consumed, a log message will be emitted when we exceed the `:buffer_size`
-  configuration.
+  configuration. This behavior can be customized by implementing the optional
+  `c:format_discarded\2` callback.
```
```suggestion
  `c:format_discarded/2` callback.
```

josevalim commented on lib/gen_stage.ex:
```diff
@@ -1693,6 +1705,14 @@ defmodule GenStage do
   end
 
+  @doc """
+  Returns the number of buffered items for a producer.
```
The documentation should be clear this is an estimate. It is not the precise amount of buffered items.

josevalim commented:
@mezz dispatch events did not use the state prior to this commit, so there was no bug before, but there was definitely a bug after your changes.

Can you please expand a bit on why it returns the state? I am worried people will try to compensate for the buffer, which is not a good idea size the buffer is not exact.

mezz commented:
I am using the state update in the test to count the total number of discarded items over time. It seemed like a simple realistic use for the callback.

The discarded items are not passed to the callback (only their count) so I am not sure what the user might try to accomplish with the state when items are discarded. If you think it is opening things to misuse I can remove the returned state, it is not necessary for my real case because I am just sending the information off to another service and not storing anything in the state.

I will update the docs and name of the method to reflect that the buffer count is just an estimate.

mezz commented:
I have removed the returned state from `format_discarded` to simplify the PR.

josevalim commented on lib/gen_stage.ex:
```diff
@@ -974,6 +975,15 @@ defmodule GenStage do
+  Invoked when items are discarded from the buffer.
+
+  It receives the number of excess (discarded) items from this invocation.
+  This callback returns a boolean that controls whether the default error log for discarded items is printed or not.
+  Return true to print the log, return false to skip the log.
```
```suggestion
  This callback returns a boolean that controls whether the default error
  log for discarded items is printed or not. It should return `true` to print
  the log, `false` otherwise.
```

josevalim commented:
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

mezz commented:
Thanks!
